performance/autoupdate-clients/main.go

```diff
@@ -0,0 +1,258 @@
+	}
+
+	go func(quit chan int) {
+		for {
+			select {
+			case <-quit:
+				return
+			case <-time.After(time.Second):
+				if !gotUpdate {
+					continue
+				}
+				gotUpdate = false
+
+				// count the amount of change ids per change id
+				amountChangeIds := make(map[int]int)
+				changeIdsMutex.Lock()
+				for _, changeId := range changeIds {
+					if _, ok := amountChangeIds[changeId]; ok {
+						amountChangeIds[changeId] += 1
+					} else {
+						amountChangeIds[changeId] = 1
+					}
+				}
+				changeIdsMutex.Unlock()
+
+				// get size of update
+				sumKB := int(sumBytes / 1024)
+				diffKB := sumKB - oldSumKB
+				oldSumKB = sumKB
+				log.Printf("New update: %d KB (diff %d KB)\n", sumKB, diffKB)
+
+				sortedChangeIds := make([]int, len(amountChangeIds))
+				i := 0
+				for changeId := range amountChangeIds {
+					sortedChangeIds[i] = changeId
+					i++
+				}
+				sort.Ints(sortedChangeIds)
+				for _, changeId := range sortedChangeIds {
+					log.Printf("%d: %d\n", changeId, amountChangeIds[changeId])
+				}
+
+				// change id delta: difference between min and max change id
+				delta := sortedChangeIds[len(sortedChangeIds)-1] - sortedChangeIds[0]
+				log.Printf("delta=%d\n", delta)
+
+				log.Printf("\n")
+			}
+		}
+	}(quit)
+
+	for {
+		select {
+		case <-interrupt:
+			for i := 0; i < (N + 1); i++ {
+				quit <- 0
+			}
+			return
+		case message := <-recv:
+			gotUpdate = true
+			sumBytes += message.Length
+			changeIdsMutex.Lock()
+			changeIds[message.I] = message.ChangeId
+			changeIdsMutex.Unlock()
+		}
+	}
+}
+
+func client(quit chan int, recv chan Message, i int, jar http.CookieJar, host string) {
+	retry := true
+	initialIteration := true
+	for retry {
+		if !initialIteration {
+			time.Sleep(1 * time.Second)
+			log.Println(i, "retry after error")
+		}
+		initialIteration = false
+
+		var client *http.Client
+		tr := &http.Transport{
+			TLSClientConfig: &tls.Config{InsecureSkipVerify: true},
+		}
+		client = &http.Client{Transport: tr, Jar: jar}
+		resp, err := client.Get(host + "/system/autoupdate")
+		if err != nil {
+			log.Println("dial: ", err)
+			continue
+		}
+
+		done := make(chan struct{})
+
+		go func(resp *http.Response, recv chan Message, done chan struct{}) {
+			defer close(done)
+			reader := bufio.NewReader(resp.Body)
+			for {
+				message, err := reader.ReadBytes('\n')
+
+				if err != nil {
+					log.Printf("Error: %v\n", err)
+					return
+				}
+
+				m := parseMessage(message, i)
+				if m != nil {
+					recv <- *m
+				}
+			}
+		}(resp, recv, done)
+
+	L:
+		for {
+			select {
+			case <-quit:
+				retry = false
+				select {
+				case <-done:
+				case <-time.After(time.Second):
+				}
+				return
+			case <-done:
+				// The recv goroutine errored. break the for and retry it
+				log.Println("recv goroutine exited!")
+				break L
+			}
+		}
+	}
+}
+
+func parseMessage(data []byte, i int) *Message {
+	var result map[string]interface{}
+	json.Unmarshal(data, &result)
```

Comment:
You should check the error.

If you are sure, that this can never fail, then you should write:
```go
_ = json.Unmarshal(...)
```
This shows, that you did not forget to check the error.